test/gtest/ucp/test_ucp_request.cc

```diff
@@ -644,87 +638,101 @@ class test_proto_reset_atp : public test_proto_reset {
     {
-        if (m_req != NULL) {
-            cb(&m_req->send.uct, arg);
-            m_req = NULL;
+        for (auto &req : m_reqs) {
+            cb(req, arg);
         }
+
+        m_reqs.clear();
     }
 
 protected:
-    void wait_and_restart(const std::vector<request_pair_t> &pairs) override
+    void wait_and_restart(const std::vector<void*> &reqs) override
     {
-        ucp_ep_h ep                = sender().ep();
-        ucp_lane_index_t num_lanes = ucp_ep_config(ep)->key.num_lanes;
-
-        /* Backup uct ops for all lanes */
-        for (ucp_lane_index_t lane = 0; lane < num_lanes; ++lane) {
-            m_ops.push_back(ucp_ep_get_lane(ep, lane)->iface->ops);
-        }
-
         hook_uct_cbs();
 
-        /* Wait for rndv_put initialization */
-        auto &pair = wait_for_condition(pairs, [](const request_pair_t &pair) {
-            std::string rndv_put_zcopy_name("rndv/put/zcopy");
-            return rndv_put_zcopy_name ==
-                   pair.sreq->send.proto_config->proto->name;
-        });
-
-        restore_uct_cbs();
+        /* Wait until first ATP was sent and received by remote side */
+        static const unsigned send_stage = 3;
+        wait_any(reqs, [reqs](const void *ureq) {
+            if (ureq == NULL) {
+                return false;
+            }
 
-        /* Wait until ATP stage starts */
-        static const unsigned send_stage = 0;
-        const double timeout             = 10;
-        ucs_time_t deadline              = ucs::get_deadline(timeout);
+            ucp_request_t *req = (ucp_request_t*)ureq - 1;
+            if (!(req->flags & UCP_REQUEST_FLAG_PROTO_SEND)) {
+                return false;
+            }
 
-        while ((pair.sreq->send.proto_stage == send_stage) &&
-               (ucs_get_time() < deadline)) {
-            pair.sreq->send.uct.func(&pair.sreq->send.uct);
-        }
+            if (req->send.proto_stage != send_stage) {
+                return false;
+            }
 
-        /* One more progress to send the first ATP */
-        pair.sreq->send.uct.func(&pair.sreq->send.uct);
+            const auto it = std::find(reqs.begin(), reqs.end(), ureq);
+            EXPECT_NE(it, reqs.end());
+            EXPECT_NE(it, reqs.end() - 1);
+            const void *user_rreq = *(it + 1);
 
-        const ucp_request_t *rndv_req = pair.rreq->recv.proto_rndv_request;
-        deadline                      = ucs::get_deadline(timeout);
+            if (user_rreq == NULL) {
+                return false;
+            }
 
-        /* Wait until receiver gets the ATP message */
-        while (rndv_req->send.state.completed_size == 0 &&
-               (ucs_get_time() < deadline)) {
-            receiver().progress();
-        }
+            ucp_request_t *rreq           = (ucp_request_t*)user_rreq - 1;
+            const ucp_request_t *rndv_req = rreq->recv.proto_rndv_request;
+            if (rndv_req == NULL) {
+                return false;
```

Comment:
why need to find the matching receive request?

Reply:
how else can wait for receiver to process it?

Reply:
i don't think we must wait for it here. can do it before the test finishes. as long as we call progress, the recv reqs will also progress